lib/services/local/users.go

```diff
@@ -441,58 +441,76 @@ func (s *IdentityService) UpsertUser(ctx context.Context, user types.User) (type
 	return user, nil
 }
 
-// CompareAndSwapUser updates a user, but fails if the value (as exists in the
-// backend) differs from the provided `existing` value. If the existing value
+// CompareAndSwapUser updates a user, but fails if the user (as exists in the
+// backend) differs from the provided `existing` user. If the existing user
 // matches, returns no error, otherwise returns `trace.CompareFailed`.
 func (s *IdentityService) CompareAndSwapUser(ctx context.Context, new, existing types.User) error {
+	if new.GetName() != existing.GetName() {
+		return trace.BadParameter("name mismatch between new and existing user")
+	}
 	if err := services.ValidateUser(new); err != nil {
 		return trace.Wrap(err)
 	}
 
-	newRaw, ok := new.WithoutSecrets().(types.User)
+	newWithoutSecrets, ok := new.WithoutSecrets().(types.User)
 	if !ok {
-		return trace.BadParameter("Invalid user type %T", new)
+		return trace.BadParameter("Invalid new user type %T (this is a bug).", new)
 	}
-	rev := new.GetRevision()
-	newValue, err := services.MarshalUser(newRaw)
-	if err != nil {
-		return trace.Wrap(err)
+
+	existingWithoutSecrets, ok := existing.WithoutSecrets().(types.User)
+	if !ok {
+		return trace.BadParameter("Invalid existing user type %T (this is a bug).", existing)
 	}
-	newItem := backend.Item{
+
+	item := backend.Item{
 		Key:      backend.Key(webPrefix, usersPrefix, new.GetName(), paramsPrefix),
-		Value:    newValue,
+		Value:    nil, // avoid marshaling new until we pass one comparison
 		Expires:  new.Expiry(),
-		ID:       new.GetResourceID(),
-		Revision: rev,
+		Revision: "",
 	}
 
-	existingRaw, ok := existing.WithoutSecrets().(types.User)
-	if !ok {
-		return trace.BadParameter("Invalid user type %T", existing)
-	}
-	existingValue, err := services.MarshalUser(existingRaw)
-	if err != nil {
-		return trace.Wrap(err)
-	}
-	existingItem := backend.Item{
-		Key:   backend.Key(webPrefix, usersPrefix, existing.GetName(), paramsPrefix),
-		Value: existingValue,
-	}
+	const iterationLimit = 5
```

Comment:
What is the rationale behind the retries? Are all ConditionalUpdate callers expected to retry? Why 5 times exactly?

Reply:
Revisions can change either spuriously (although this doesn't actually in any of our backends right now, outside of insane situations like restoring a backup while the auth server is still running) or because something else modified the item - but from the point of view of a compare-and-swap operation we don't care about that, we only care if the value matches or not, and a concurrent modification that leaves the value identical should still result in a successful compare-and-swap. In practice this isn't actually verified in some of our `CompareAndSwapFoo` implementations, unfortunately: `CompareAndSwapCertAuthority` can fail even though it should succeed if the existing value is concurrently overwritten with a different one that is equivalent after deserialization (in that specific case it's not a big deal because all cases of CASing a cert authority are already on various timed loops to begin with).

tl;dr: conditional updates should always be treated as potentially fallible, and the behavior on a failure should be that you reread the current state again and then make a decision on whether you can and want to keep going, or return an error and let the outer layer deal with that

Oh, and I picked 5 because humans generally have 5 fingers on one hand, although this is such a fringe scenario that I'm going to change it to 3: the initial attempt, one retry because spurious failures could actually happen, and another one because a single retry just feels weird.

Reply:
> and I picked 5 because humans generally have 5 fingers on one hand (...)

That's rock-solid logic, can't argue with that.

Thanks for the explanation. So what I'm gathering is that looped ConditionalUpdates are a practice we'd like to encourage, which is good to know.

Do you think it makes sense to capture the number of retries in a more generic, reusable constant, like `ConditionalUpdateDefaultRetries` (or something to that effect)?

Reply:
I don't know if there's a one-size-fits-all value for such a constant, I'm only doing a retry loop very close to the backend here because I want to match the slightly stronger semantics of compare-and-swap.